internal/setup/setup.go

```diff
@@ -158,35 +193,63 @@ func (r *Release) validate() error {
 						return fmt.Errorf("slices %s and %s conflict on %s", old, new, newPath)
 					}
 				} else {
-					if newInfo.Kind == GlobPath {
-						globs[newPath] = new
+					switch newInfo.Kind {
+					case GlobPath:
+						globs = append(globs, pathSlice{path: newPath, slice: new})
+					case CopyPath:
+						copies = append(copies, pathSlice{path: newPath, slice: new})
+					case GeneratePath:
+						generates = append(generates, pathSlice{path: newPath, slice: new})
+					default:
+						rest = append(rest, pathSlice{path: newPath, slice: new})
 					}
-					paths[newPath] = new
+					allPaths[newPath] = new
 				}
 			}
 		}
 	}
 
-	// Check for cycles.
-	_, err := order(r.Packages, keys)
-	if err != nil {
-		return err
+	checkConflict := func(old, new pathSlice) error {
+		if strdist.GlobPath(new.path, old.path) {
+			if (old.slice.Package > new.slice.Package) || (old.slice.Package == new.slice.Package && old.slice.Name > new.slice.Name) {
+				old, new = new, old
+			}
+			return fmt.Errorf("slices %s and %s conflict on %s and %s", old.slice, new.slice, old.path, new.path)
+		}
+		return nil
 	}
-
-	// Check for glob conflicts.
-	for newPath, new := range globs {
-		for oldPath, old := range paths {
-			if new.Package == old.Package {
+	for _, new := range globs {
+		// TODO replace with slices.Concat once we upgrade to go 1.22.
+		for _, old := range append(globs, copies...) {
+			if new.slice.Package == old.slice.Package {
```

Comment:
This change introduced a new type (pathSlice), a new closure (checkConflict), four new slices (rest, copies, generates, globs), copy & pastes the new checkConflict(old, new) logic three different times, duplicates a N^N loop, not to mention it re-appends a large slice to another large slice _every iteration of the loop._ Indeed does that to _three_ slices on the other case below.

This is not an improvement. This logic needs tuning and I can see that you understand why, but can we please go back to the original state and fix the problem there?

Reply:
Agree, it was very messy and did a lot of copying information around that we were essentially duplicating. I was struggling to find a way to organize it so that it is clear when reading the code, which was something that did not happen in the past, but now, between the new comments and the new logic, I think I am finally happy that it is good enough (it will never be perfect). Please look at the PR diff and tell me what you think, there is still room for performance improvements, that is for sure, but I don't think it is necessary now.